// figma-calculations/dist/rules/cleaners/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LintCleaner = exports.removeEmptyFramesAndGroups = exports.removeHiddenNodes = exports.removeTransparentNodes = exports.removeNodesOutsideFrameBounds = void 0;
const removeNodesOutsideFrameBounds_1 = require("./removeNodesOutsideFrameBounds");
Object.defineProperty(exports, "removeNodesOutsideFrameBounds", { enumerable: true, get: function () { return removeNodesOutsideFrameBounds_1.removeNodesOutsideFrameBounds; } });
const removeTransparentNodes_1 = require("./removeTransparentNodes");
Object.defineProperty(exports, "removeTransparentNodes", { enumerable: true, get: function () { return removeTransparentNodes_1.removeTransparentNodes; } });
const removeEmptyFramesAndGroups_1 = require("./removeEmptyFramesAndGroups");
Object.defineProperty(exports, "removeEmptyFramesAndGroups", { enumerable: true, get: function () { return removeEmptyFramesAndGroups_1.removeEmptyFramesAndGroups; } });
const removeHiddenNodes_1 = require("./removeHiddenNodes");
Object.defineProperty(exports, "removeHiddenNodes", { enumerable: true, get: function () { return removeHiddenNodes_1.removeHiddenNodes; } });
class LintCleaner {
    static run(startNode) {
        (0, removeEmptyFramesAndGroups_1.removeEmptyFramesAndGroups)(startNode);
        (0, removeTransparentNodes_1.removeTransparentNodes)(startNode);
        // removeHiddenNodes(startNode);
        // removeNodesOutsideFrameBounds(startNode);
    }
}
exports.LintCleaner = LintCleaner;
exports.default = LintCleaner;
